azure-spring-maven-plugin/src/main/java/com/microsoft/azure/maven/spring/pom/PomXmlUpdater.java

```diff
@@ -0,0 +1,242 @@
+import org.dom4j.QName;
+import org.dom4j.Text;
+import org.dom4j.dom.DOMElement;
+import org.dom4j.io.SAXContentHandler;
+import org.dom4j.io.SAXReader;
+import org.dom4j.tree.DefaultElement;
+import org.dom4j.tree.DefaultText;
+import org.xml.sax.Locator;
+import org.xml.sax.XMLReader;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.util.Arrays;
+import java.util.stream.Collectors;
+
+public class PomXmlUpdater {
+    private MavenProject project;
+    private PluginDescriptor plugin;
+
+    public PomXmlUpdater(MavenProject project, PluginDescriptor plugin) {
+        this.project = project;
+        this.plugin = plugin;
+    }
+
+    public void updateSettings(AppSettings app, DeploymentSettings deploy) throws DocumentException, IOException {
+        final File pom = this.project.getFile();
+        final SAXReader reader = new CustomSAXReader();
+        reader.setDocumentFactory(new LocatorAwareDocumentFactory());
+        final Document doc = reader.read(new InputStreamReader(new FileInputStream(pom)));
+        final Element pluginsNode = createToPath(doc.getRootElement(), "build", "plugins");
+        Element springPluginNode = null;
+        for (final Element element : pluginsNode.elements()) {
+            final String groupId = XmlUtils.getChildValue("groupId", element);
+            final String artifactId = XmlUtils.getChildValue("artifactId", element);
+            final String version = XmlUtils.getChildValue("version", element);
+            if (plugin.getGroupId().equals(groupId) && plugin.getArtifactId().equals(artifactId) &&
+                    (plugin.getVersion().equals(version) || version == null)) {
+                springPluginNode = element;
+            }
+        }
+        if (springPluginNode == null) {
+            springPluginNode = createMavenSpringPluginNode(pluginsNode);
+        }
+
+        final Element configurationNode = createToPath(springPluginNode, "configuration");
+        app.applyToDom4j(configurationNode);
+        final Element deployNode = createToPath(configurationNode, "deployment");
+        deploy.applyToDom4j(deployNode);
+        // newly created nodes are not LocationAwareElement
+        // use configurationNode as initial value since we want to format configuration node if it exists.
+        Element newNode = configurationNode;
+        while (!(newNode.getParent() instanceof LocationAwareElement)) {
+            newNode = newNode.getParent();
+        }
+        FileUtils.fileWrite(pom, formatElement((LocationAwareElement) newNode.getParent(), newNode));
+    }
+
+    private static String formatElement(LocationAwareElement parent, Element newNode) {
+        final String[] originXmlLines = IndentUtil.splitLines(parent.getDocument().asXML());
+        final String baseIndent = IndentUtil.calcXmlIndent(originXmlLines, parent.getLineNumber() - 1,
+                parent.getColumnNumber() - 2);
```

Comment:
`parent.getLineNumber() - 1`, `parent.getColumnNumber() - 2`, could you add some comments about the calculation of line/col number ?

Reply:
columnNumber is the next char after tag name, so here I minus 2